### PR TITLE
forward data-* attributes to t-component root node

### PR DESCRIPTION
**Pull request by brboi**

### CODE REVIEW

File: src/component/directive.ts [hunk 1]
```diff
@@ -271,13 +271,39 @@ QWeb.addDirective({
     let tattClass = node.getAttribute("t-att-class");
     let styleAttr = node.getAttribute("style");
     let tattStyle = node.getAttribute("t-att-style");
+
+    const dataAttrs = [...node.attributes].filter(a => /^(t-attf?-)?data-.+$/.test(a.name)).sort();
+    if (dataAttrs.length) {
+      const dashToCamel = (str: string) => {
```
Comment (aab-odoo) [inline src/component/directive.ts:277]: dash style is called kebab-case, so maybe you want to rename into `kebabToCamel` ?

File: src/component/directive.ts [hunk 2]
```diff
@@ -271,13 +271,39 @@ QWeb.addDirective({
     let tattClass = node.getAttribute("t-att-class");
     let styleAttr = node.getAttribute("style");
     let tattStyle = node.getAttribute("t-att-style");
+
+    const dataAttrs = [...node.attributes].filter(a => /^(t-attf?-)?data-.+$/.test(a.name)).sort();
```
Comment (aab-odoo) [inline src/component/directive.ts:275]: sort?

File: src/component/directive.ts [hunk 3]
```diff
@@ -271,13 +271,39 @@ QWeb.addDirective({
...
+    if (dataAttrs.length) {
+      const dashToCamel = (str: string) => {
+        return str.toLowerCase()
+          .replace(/(\-[a-z])/g, val => val.toUpperCase().replace('-', ''));
+      };
+      ctx.addLine("let dataAttrMap = {};")
```
Comment (aab-odoo) [inline src/component/directive.ts:281]: could be `const`

File: src/component/directive.ts [hunk 4]
```diff
@@ -271,13 +271,39 @@ QWeb.addDirective({
...
+      let attrName = "", attrValue = "";
```
Comment (aab-odoo) [inline src/component/directive.ts:282]: typically, we try not to do that, and rather define one variable per line (full definition: `let machin = "val";`)

File: src/component/directive.ts [hunk 5]
```diff
@@ -271,13 +271,39 @@ QWeb.addDirective({
...
+      for (const dataAttr of dataAttrs) {
+        if (dataAttr.name.startsWith("t-att-")) {
+          attrName = dashToCamel(dataAttr.name.slice(11));
+          let v = ctx.getValue(dataAttr.value);
+          attrValue = typeof v === "string" ? ctx.formatExpression(v) : `scope.${v.id}`;
```
Comment (aab-odoo) [inline src/component/directive.ts:287]: could this not be a string?

File: src/component/directive.ts [hunk 6]
```diff
@@ -271,13 +271,39 @@ QWeb.addDirective({
...
+        } else if (dataAttr.name.startsWith("t-attf-")) {
+          attrName = dashToCamel(dataAttr.name.slice(12));
+          let v = ctx.getValue(dataAttr.value);
+          attrValue = typeof v === "string" ? ctx.interpolate(v) : `scope.${v.id}`;
```
Comment (aab-odoo) [inline src/component/directive.ts:291]: same

File: src/component/directive.ts [hunk 7]
```diff
@@ -271,13 +271,39 @@ QWeb.addDirective({
...
+        } else {
+          attrName = dashToCamel(dataAttr.name.slice(5));
+          attrValue = '"' + dataAttr.value + '"';
```
Comment (aab-odoo) [inline src/component/directive.ts:294]: maybe use template string

File: src/component/directive.ts [hunk 8]
```diff
@@ -271,13 +271,39 @@ QWeb.addDirective({
...
+        }
+        ctx.addLine(`dataAttrMap.${attrName} = dataAttrMap.${attrName} ? dataAttrMap.${attrName} + " " + ${attrValue} : ${attrValue};`)
```
Comment (aab-odoo) [inline src/component/directive.ts:296]: Not sure it makes sense to combine (by concatenation) data values. It works well for classnames, but here I guess I wouldn't do that.

**Review by aab-odoo — COMMENTED** [review]
As we discussed, it's up to @ged-odoo to decide whether or not we want this. But in any case, good effort because this code isn't easy to dive into.

File: src/component/directive.ts (hunk 2 above)
Comment (brboi) [reply src/component/directive.ts]: This needs to be discussed.
What to do when template looks like this:
```xml
<Component t-att-data-foo="var" data-foo="42" t-attf-data-foo="var is {{var}}" />
```

Sorting is not the solution, though :-D

File: src/component/directive.ts (hunk 5 above)
Comment (brboi) [reply src/component/directive.ts:289]: I need to triple check, but I think this could be a QWebVar too

File: src/component/directive.ts (hunk 8 above)
Comment (brboi) [reply src/component/directive.ts]: I agree. We should probably log warnings when template define more than once the same data-* attribute.

File: src/component/directive.ts (hunk 5 above)
Comment (brboi) [reply src/component/directive.ts]: You were right.

**Comment by LuvForAirplanes** [comment]
This was fixed in #903. Thanks for the effort.

**Pull request closed without merging**